### PR TITLE
Feature/websockets real timing

### PR DESCRIPTION
**Pull request by trinatafoya**
This MR handles websockets with message timing.

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -4,12 +4,12 @@
 
 "@ava/babel-plugin-throws-helper@^3.0.0":
   version "3.0.0"
-  resolved "https://registry.yarnpkg.com/@ava/babel-plugin-throws-helper/-/babel-plugin-throws-helper-3.0.0.tgz#2c933ec22da0c4ce1fc5369f2b95452c70420586"
+  resolved "https://registry.npmjs.org/@ava/babel-plugin-throws-helper/-/babel-plugin-throws-helper-3.0.0.tgz"
```
Comment (NeedleInAJayStack) [inline yarn.lock]: Any reason why there are so many changes to this file? Perhaps we used `npm install` instead of `yarn install`? Unless these yarn changes are required for the change, I would remove them.

File: .husky/pre-commit [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+npm test
```
Comment (NeedleInAJayStack) [inline .husky/pre-commit]:
```suggestion
yarn test
```

File: bin/har-to-k6.js [hunk 3]
```diff
@@ -96,9 +96,8 @@ function success(output, log) {
 function inform(error, log) {
   log.error(`${summarize(error)}:`)
   const source = cause(error)
-  const message = `${source.message}${
-    source.path ? `, at path: ${source.path}` : ''
-  }`
+  const message = `${source.message}${source.path ? `, at path: ${source.path}` : ''
+    }`
```
Comment (NeedleInAJayStack) [inline bin/har-to-k6.js]: nit: Undo this change

File: src/parse/entry.js [hunk 4]
```diff
@@ -23,6 +23,13 @@ function entry(node, result) {
   if (node.variables) {
     variables(node.variables, spec.variables)
   }
+  if (node._webSocketMessages) {
+    // capture all the websocket messages for later use connecting to the websocket
+    let secondsFromMilisecondsMultiplier = 0.001
+    spec.timeConnected = node.time * secondsFromMilisecondsMultiplier
```
Comment (NeedleInAJayStack) [inline src/parse/entry.js:29]: Great use of variable names to clarify the code!

File: src/parse/imports.js [hunk 5]
```diff
@@ -42,9 +42,21 @@ function imports(archive, result) {
...
+      result.imports.websocket = true
+    }
   }
 }
 
+function webSocketItem(entries) {
+  // This function checkes to see if the request is a ws or wss
+  //  request to make sure imports has websocket true
+  const check = ({ request }) =>
+    ['ws:', 'wss'].includes(request.url.slice(0, 3))
```
Comment (NeedleInAJayStack) [inline src/parse/imports.js]: nit: Since we do this twice, would it make sense to make an `isWebsocket` function on `request`?

File: src/render/websocket.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  let messagesWithNewTimes = JSON.parse(JSON.stringify(messages))
+  messagesWithNewTimes[firstIndexSetToZero].time = firstTimeStartsAtZero
+  for (let i = 1; i < messages.length; i++){
+    messagesWithNewTimes[i].time = messages[i].time - messages[i - 1].time
+  }
+  return messagesWithNewTimes
+}
+
+function timeAlive(timeConnected, messages) {
+  /// calculates length of time when websocket is open minus the total time the messages took to run before closeing the socket
```
Comment (NeedleInAJayStack) [inline src/render/websocket.js]:
```suggestion
  /// calculates length of time when websocket is open minus the total time the messages took to run before closing the socket
```

File: src/render/websocket.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  let firstMessageTime = messages[0].time
+  let finalMessageTime = messages[messages.length - 1].time
```
Comment (NeedleInAJayStack) [inline src/render/websocket.js]: nit: Consider checking that messages has at least 1 item

File: src/render/websocket.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    )
+  }
+
+  send_messages.push(
+    ...[
+      'socket.close()',
+      '})',
+      "socket.on('error', function (e) {",
+      'fail(`WebSocket failed: ${e.error()}`);',
+      '})\n}\n',
```
Comment (NeedleInAJayStack) [inline src/render/websocket.js]: nit: Split these onto separate strings and let the join add in newlines

**Review by NeedleInAJayStack — COMMENTED** [review]
Nice work!

Could we add some automated unit tests to ensure that this continues to work as intended?

File: src/render/websocket.js (hunk 7 above)
Comment (trinatafoya) [reply src/render/websocket.js]: Good IDEA i also will change function changeMessageTimes on line 32

File: src/render/websocket.js (hunk 8 above)
Comment (trinatafoya) [reply src/render/websocket.js]: Okay

File: src/render/websocket.js (hunk 6 above)
Comment (trinatafoya) [reply src/render/websocket.js]: Ok

File: src/parse/imports.js (hunk 5 above)
Comment (trinatafoya) [reply src/parse/imports.js]: Ok

**Comment by NeedleInAJayStack** [comment]
Great job!!

**Review by NeedleInAJayStack — APPROVED** [review]
